docs/architecture/adr-001-flush-version.md

```diff
@@ -0,0 +1,111 @@
+# ADR 001: Flush Version
+
+## Changelog
+
+- May 22, 2020: Initial Draft
+
+## Status
+
+Proposed
+
+## Context
+
+The IAVL library recently underwent changes to the pruning and version commitment model. Specifically,
+the invariant that a version is flushed to disk when it is committed via `SaveVersion` no longer holds
+true. Instead, versions are kept in memory and periodically flushed to disk and then pruned based on
+the client supplied pruning strategy parameters. For more detailed information, see the
+[PRUNING](../tree/PRUNING.md) document.
+
+These changes, while drastically improving performance under certain circumstances, introduces certain
+tradeoffs. Specifically, an application with a deterministic state-machine that commits and merkle-izes
+state via IAVL tree(s), can no longer guarantee that a state is written to disk when it is committed
+and must rely on some sort of state replay mechanism (e.g. Tendermint peer-to-peer gossip & state machine execution). While this inherently
+is not necessarily a problem, it becomes a problem under certain contexts that depend on a specific
+version existing on disk.
+
+One such example is live upgrades. Specifically, when a live upgrade occurs in some application, typically
+a new binary will be started in place of the old binary and some set of business logic will need to
+be executed in order to migrate and handle old and/or invalid state to make it compatible with the
+new version. Being that these operations must occur on the latest canonical state, we must ensure the
+latest version is committed and flushed to disk so all operators and clients see the same view when
+doing this upgrade. Hence, we need a means to manually signal to the IAVL tree that a specific
+version should be flushed to disk. In addition, we also need to ensure this version is not pruned at
+a later point in time.
+
+## Decision
+
+We propose to introduce a new API, `FlushVersion(version int64)`. This new API will attempt to manually
+flush a previously committed version to disk. It will return an error if the version does not exist
+or if the underlying flush fails. If the version is already flushed, the method performs a no-op
+and no error is returned. After the version is flushed to disk and the version is not the latest
+version, we will also remove the version from the recent (in-memory) database via `DeleteVersionFromRecent`.
+
+In order to not have a manually flushed version pruned at a later point in time
+(e.g. after an application restarts), we also introduce a new type, `CommitMetadata`, that will be
+flushed to disk for every version saved, regardless if it's flushed to disk.
```

Comment:
What is the benefit of always flushing this to disk? Will metadata be kept even for deleted versions? What about future versions that were discarded via `LoadVersionForOverwriting`? Should there be an API to fetch and delete metadata?

Reply:
> What is the benefit of always flushing this to disk?

It makes the model easier to reason about. Alternatively, we can flush to disk periodically. Most importantly, we need to think about any edge cases where a client may halt and a subset of metadata was not flushed to disk. What happens in this case when the metadata needs to be fetched and isn't there?

> Will metadata be kept even for deleted versions?

In this ADR, that is the implicit assumption, yes. Do you propose we delete metadata along with the version when the version is deleted?

> What about future versions that were discarded via `LoadVersionForOverwriting`?

This would overwrite any metadata that exists for an already committed version.

> Should there be an API to fetch and delete metadata?

There will be an API to retrieve metadata, but not delete it. Metadata should only be deleted as the internal API sees fit.

Reply:
> we need to think about any edge cases where a client may halt and a subset of metadata was not flushed to disk.

We could do two updates: store the metadata with a flag saying that the version is about to be saved, and then flip the flag once it has been saved. That would allow us to detect incomplete saves as well.

> Do you propose we delete metadata along with the version when the version is deleted?

No, I think it probably makes sense to keep metadata around anyway. But we would have to handle backwards compatibility with existing applications, and handle the missing metadata.

Reply:
> We could do two updates: store the metadata with a flag saying that the version is about to be saved, and then flip the flag once it has been saved. That would allow us to detect incomplete saves as well.

Could you elaborate a bit? Store the metadata where? Do you think it'll be easier to just persist to disk every save? It'll only be a few bytes.

Reply:
> Could you elaborate a bit?

I mean, when we call `SaveVersion()` we would do the following:

* Store `CommitMetadata{Version: 10, Status: "saving"}` to disk
* Store the version to memory or disk as appropriate
* Store `CommitMetadata{Version: 10, Status: "saved"}` to disk

> Store the metadata where? Do you think it'll be easier to just persist to disk every save? It'll only be a few bytes.

Yeah, we should always keep the metadata on disk.

Reply:
I see. What does this buy us as opposed to just saving metadata to disk once? It can even be part of the snapshot batch to make it atomic.

Reply:
This was in response to your original comment

> we need to think about any edge cases where a client may halt and a subset of metadata was not flushed to disk

I don't think tm-db guarantee that batches are always transactional or atomic, as it depends on the backend. I believe some of them simply implement batches as a foreach.